qutebrowser/commands/runners.py

```diff
@@ -49,21 +49,29 @@ def _current_url(tabbed_browser):
 
 def replace_variables(win_id, arglist):
     """Utility function to replace variables like {url} in a list of args."""
+    variables = {
+        '{url}': lambda: _current_url(tabbed_browser).toString(
+            QUrl.FullyEncoded | QUrl.RemovePassword),
+        '{url:pretty}': lambda: _current_url(tabbed_browser).toString(
+            QUrl.RemovePassword),
+        '{clipboard}': utils.get_clipboard,
+        '{primary}': lambda: utils.get_clipboard(selection=True),
+    }
+    values = {}
     args = []
     tabbed_browser = objreg.get('tabbed-browser', scope='window',
                                 window=win_id)
-    if any('{url}' in arg for arg in arglist):
-        url = _current_url(tabbed_browser).toString(QUrl.FullyEncoded |
-                                                    QUrl.RemovePassword)
-    if any('{url:pretty}' in arg for arg in arglist):
-        pretty_url = _current_url(tabbed_browser).toString(QUrl.RemovePassword)
-    for arg in arglist:
-        if '{url}' in arg:
-            args.append(arg.replace('{url}', url))
-        elif '{url:pretty}' in arg:
-            args.append(arg.replace('{url:pretty}', pretty_url))
-        else:
+
+    try:
+        for arg in arglist:
+            for var, func in variables.items():
+                if var in arg:
+                    if var not in values:
+                        values[var] = func()
```

Comment:
Why do you cycle over the dictionary? If you are concerned about the possibility of receiving a variable that is not defined a simple test like this would work:

```python
for arg in arglist:
  if arg in variables:
    if var not in values:
      values[var] = variables[var]()
  else:
    raise WhateverException('{} is not an allowed variable placeholder'.format(var))
```